File: 프로그래머스/1/172928. 공원 산책/공원 산책.py
def solution(park, routes):
    # 좌표 변수 선언
    x,y=0,0
    # 방향 해시 코드
    d={'E':[0,1],'N':[-1,0],'S':[1,0],'W':[0,-1]}
    # 시작 좌표 찾기
    for i,s in enumerate(park):
        if s.find('S')!=-1:
            x=i
            y=s.find('S')
            break
    # 수행할 명령 리스트를 돌면서
    for w in routes:
        i,m=w.split()
        # 벗어난 범위 또는 장애물 플래그 변수 선언
        obs=False
        dx=d[i][0]*int(m)
        dy=d[i][1]*int(m)
        # 범위를 벗어났을 때
        if x+dx<0 or x+dx>=len(park) or y+dy<0 or y+dy>=len(park[0]):
            obs=True
        else:
            # 명령의 방향이 남 또는 북쪽일 때
            if i=='S' or i=='N':
                # 명령 방향으로 이동하면서 장애물 찾기
                for k in range(x,x+dx+d[i][0],d[i][0]):
                    if park[k][y]=='X':
                        obs=True
                        break
                # 장애물이 없을 때
                if not obs:
                    x+=dx
            # 명령의 방향이 동 또는 서쪽일 때
            else: # elif i=='E' or i=='W':
                # 명령 방향으로 이동하면서 장애물 찾기
                for j in range(y,y+dy+d[i][1],d[i][1]):
                    if park[x][j]=='X':
                        obs=True
                        break
                # 장애물이 없을 때
                if not obs:
                    y+=dy
    r=[x,y]
    return r

File: 프로그래머스/1/172928. 공원 산책/공원 산책.cpp
#include <string>
#include <vector>
#include <map>
#include <sstream>

using namespace std;

vector<int> solution(vector<string> park, vector<string> routes) {
    // 좌표 변수 선언
    int x=0,y=0,m,dx,dy;
    char c;
    bool obs;
    // 방향 해시 코드
    map<char,vector<int>> d={{'E',{0,1}},{'S',{1,0}},{'W',{0,-1}},{'N',{-1,0}}};
    // 시작 좌표 찾기
    for(int i=0;i<park.size();i++) {
        if(park[i].find("S")!=string::npos) {
            x=i;
            y=park[i].find("S");
            break;
        }
    }
    // 수행 명령 리스트 돌면서
    for(int i=0;i<routes.size();i++) {
        // 문자열 공백 기준으로 분리해 문자 유형에 맞게 삽입
        istringstream iss(routes[i]);
        iss>>c>>m;
        // 벗어난 범위 또는 장애물 플래그 변수 선언
        obs=false;
        dx=d[c][0]*m;
        dy=d[c][1]*m;
        // 범위를 벗어났을 때
        if(x+dx<0 or x+dx>=park.size() or y+dy<0 or y+dy>=park[0].size())
            obs=true;
        // 범위를 벗어나지 않았을 때
        else {
            // 명령의 방향이 남쪽 또는 북쪽일 때
            if(c=='S' or c=='N') {
                // 명령 방향 이동하면서 장애물 찾기
                if(d[c][0]>0) {
                    for(int k=x;k<x+dx+d[c][0];k+=d[c][0]) {
                        if(park[k][y]=='X') {
                            obs=true;
                            break;
                        }
                    }
                } else {
                    for(int k=x;k>x+dx+d[c][0];k+=d[c][0]) {
                        if(park[k][y]=='X') {
                            obs=true;
                            break;
                        }
                    }
                }
                // 장애물이 없으면 값 추가
                if(obs==false)
                    x+=dx;
            }
            // 명령의 방향이 동쪽 또는 서쪽일 때
            else {
                // 명령 방향 이동하면서 장애물 찾기
                if(d[c][1]>0) {
                    for(int k=y;k<y+dy+d[c][1];k+=d[c][1]) {
                        if(park[x][k]=='X') {
                            obs=true;
                            break;
                        }
                    }
                } else {
                    for(int k=y;k>y+dy+d[c][1];k+=d[c][1]) {
                        if(park[x][k]=='X') {
                            obs=true;
                            break;
                        }
                    }
                }
                // 장애물이 없으면 값 추가
                if(obs==false)
                    y+=dy;
            }
        }
    }
    vector<int> r={x,y};
    return r;
}